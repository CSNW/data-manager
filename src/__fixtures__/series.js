export function single() {
  return [
    {
      values: [
        { a: 0, b: 10 },
        { a: 1, b: 8 },
        { a: 2, b: 6 },
        { a: 3, b: 4 },
        { a: 4, b: 2 },
        { a: 5, b: 0 }
      ]
    }
  ];
}

export function multi() {
  return [single()[0], single()[0], single()[0]];
}
