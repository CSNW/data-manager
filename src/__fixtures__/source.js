import Source from '../source';
import { single } from './series';

export default function source() {
  const override = new Source();
  override.fetch = jest.fn(() => Promise.resolve(single()[0].values));

  return override;
}
