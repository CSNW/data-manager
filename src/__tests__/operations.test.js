import {
  filter,
  map,
  flatMap,
  sort,
  sortBy,
  groupBy,
  clone
} from '../operations';
import * as series from '../../tests/__fixtures__/series';

describe('filter', () => {
  test('should filter single series', () => {
    expect(filter(row => row.a > 2)(series.single())).toMatchSnapshot();
  });
  test('should filter multi series', () => {
    expect(filter(row => row.a <= 2)(series.multi())).toMatchSnapshot();
  });
});

describe('map', () => {
  const mapper = row => {
    row.c = row.a + row.b;
    return row;
  };

  test('should map single series', () => {
    expect(map(mapper)(series.single())).toMatchSnapshot();
  });
  test('should map multi series', () => {
    expect(map(mapper)(series.multi())).toMatchSnapshot();
  });
  test('should mutate by default', () => {
    const original = series.single();
    const mapped = map(mapper)(original);

    expect(original[0].values[0]).toBe(mapped[0].values[0]);
  });
});

describe('flatMap', () => {
  const mapper = row => {
    return [
      { a: row.a, b: row.b, c: row.a + row.b },
      { a: row.a, b: row.b, c: row.a - row.b }
    ];
  };

  test('should flatMap single series', () => {
    expect(flatMap(mapper)(series.single())).toMatchSnapshot();
  });
  test('should flatMap multi series', () => {
    expect(flatMap(mapper)(series.multi())).toMatchSnapshot();
  });
});

describe('sort', () => {
  test('should sort single series', () => {
    expect(sort((a, b) => a.b - b.b)(series.single())).toMatchSnapshot();
  });
  test('should sort multi series', () => {
    expect(sort((a, b) => b.a - a.a)(series.multi())).toMatchSnapshot();
  });
});

describe('sortBy', () => {
  test('should sortBy single series', () => {
    expect(sortBy('b', (a, b) => a - b)(series.single())).toMatchSnapshot();
  });
  test('should sortBy multi series', () => {
    expect(sortBy('a', (a, b) => b - a)(series.multi())).toMatchSnapshot();
  });
});

describe('clone', () => {
  test('should clone row objects', () => {
    const original = series.single();
    const cloned = clone()(original);

    expect(original[0].values[0].a).toEqual(cloned[0].values[0].a);
    expect(original[0].values[0]).not.toBe(cloned[0].values[0]);
  });
});

describe('groupBy', () => {
  test('should work', () => {
    expect(groupBy()).toBeDefined();
  });
});
