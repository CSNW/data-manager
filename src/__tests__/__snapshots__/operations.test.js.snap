// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter should filter multi series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 2,
        "b": 6,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 2,
        "b": 6,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 2,
        "b": 6,
      },
    ],
  },
]
`;

exports[`filter should filter single series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 5,
        "b": 0,
      },
    ],
  },
]
`;

exports[`map should filter multi series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
        "c": -10,
      },
      Object {
        "a": 1,
        "b": 8,
        "c": -7,
      },
      Object {
        "a": 2,
        "b": 6,
        "c": -4,
      },
      Object {
        "a": 3,
        "b": 4,
        "c": -1,
      },
      Object {
        "a": 4,
        "b": 2,
        "c": 2,
      },
      Object {
        "a": 5,
        "b": 0,
        "c": 5,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
        "c": -10,
      },
      Object {
        "a": 1,
        "b": 8,
        "c": -7,
      },
      Object {
        "a": 2,
        "b": 6,
        "c": -4,
      },
      Object {
        "a": 3,
        "b": 4,
        "c": -1,
      },
      Object {
        "a": 4,
        "b": 2,
        "c": 2,
      },
      Object {
        "a": 5,
        "b": 0,
        "c": 5,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
        "c": -10,
      },
      Object {
        "a": 1,
        "b": 8,
        "c": -7,
      },
      Object {
        "a": 2,
        "b": 6,
        "c": -4,
      },
      Object {
        "a": 3,
        "b": 4,
        "c": -1,
      },
      Object {
        "a": 4,
        "b": 2,
        "c": 2,
      },
      Object {
        "a": 5,
        "b": 0,
        "c": 5,
      },
    ],
  },
]
`;

exports[`map should filter single series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 0,
        "b": 10,
        "c": 10,
      },
      Object {
        "a": 1,
        "b": 8,
        "c": 9,
      },
      Object {
        "a": 2,
        "b": 6,
        "c": 8,
      },
      Object {
        "a": 3,
        "b": 4,
        "c": 7,
      },
      Object {
        "a": 4,
        "b": 2,
        "c": 6,
      },
      Object {
        "a": 5,
        "b": 0,
        "c": 5,
      },
    ],
  },
]
`;

exports[`sort should sort multi series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
]
`;

exports[`sort should sort single series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
]
`;

exports[`sortBy should sortBy multi series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
]
`;

exports[`sortBy should sortBy single series 1`] = `
Array [
  Object {
    "values": Array [
      Object {
        "a": 5,
        "b": 0,
      },
      Object {
        "a": 4,
        "b": 2,
      },
      Object {
        "a": 3,
        "b": 4,
      },
      Object {
        "a": 2,
        "b": 6,
      },
      Object {
        "a": 1,
        "b": 8,
      },
      Object {
        "a": 0,
        "b": 10,
      },
    ],
  },
]
`;
